compiler/rustc_mir_transform/src/shim.rs

```diff
@@ -1044,6 +1049,559 @@ fn build_fn_ptr_addr_shim<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId, self_ty: Ty<'t
+                        surface_async_drop_in_place_fn,
+                        [self_ty.into()],
+                        span,
+                    ),
+                    args: vec![respan(span, Operand::Move(self_ptr.into()))],
+                    destination: return_local.into(),
+                    target: Some(return_bb),
+                    unwind: UnwindAction::Continue,
+                    call_source: CallSource::Misc,
+                    fn_span: span,
+                },
+            });
+        }
+
+        GlueStrategy::Slice { elem_ty, is_array } => {
+            let slice_async_destructor_fn =
+                tcx.require_lang_item(LangItem::SliceAsyncDestructorCtor, Some(span));
+
+            blocks = IndexVec::from_elem_n(BasicBlockData::new(None), 2);
+            let slice_async_destructor_bb = BasicBlock::new(0);
+            return_bb = BasicBlock::new(1);
+
+            let slice_ptr = if is_array {
+                let slice_ptr_ty = Ty::new_mut_ptr(tcx, Ty::new_slice(tcx, elem_ty));
+                let slice_ptr = locals.push(LocalDecl::with_source_info(slice_ptr_ty, source_info));
+
+                blocks[slice_async_destructor_bb].statements = vec![
+                    Statement { source_info, kind: StatementKind::StorageLive(slice_ptr) },
+                    Statement {
+                        source_info,
+                        kind: StatementKind::Assign(Box::new((
+                            slice_ptr.into(),
+                            Rvalue::Cast(
+                                CastKind::PointerCoercion(PointerCoercion::Unsize),
+                                Operand::Move(self_ptr.into()),
+                                slice_ptr_ty,
+                            ),
+                        ))),
+                    },
+                ];
+                slice_ptr
+            } else {
+                self_ptr
+            };
+
+            blocks[slice_async_destructor_bb].terminator = Some(Terminator {
+                source_info,
+                kind: TerminatorKind::Call {
+                    func: Operand::function_handle(
+                        tcx,
+                        slice_async_destructor_fn,
+                        [elem_ty.into(), /* HOST */ tcx.consts.false_.into()],
+                        span,
+                    ),
+                    args: vec![respan(span, Operand::Move(slice_ptr.into()))],
+                    destination: return_local.into(),
+                    target: Some(return_bb),
+                    unwind: UnwindAction::Continue,
+                    call_source: CallSource::Misc,
+                    fn_span: span,
+                },
+            });
+
+            if is_array {
+                blocks[return_bb].statements =
+                    vec![Statement { source_info, kind: StatementKind::StorageDead(slice_ptr) }];
+            }
+        }
+
+        GlueStrategy::Chain { has_surface_async_drop, elem_tys } => {
```

Comment:
can we write most of this one in surface Rust somehow? Maybe we could generate a tuple of the various types that should be dropped and recursively invoke drop on them?

Reply:
Sorry but I don't understand what you are describing. If you imagine a handwritten async destructor future for a tuple it would be an enum, for which each variant of it represents running an async destructor for some field, sequentially ordered one after the other + variants like `Unresume` and `Done`.